src/librustc_typeck/coherence/builtin.rs

```diff
@@ -0,0 +1,349 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Check properties that are required by built-in traits and set
+//! up data structures required by type-checking/translation.
+
+use rustc::middle::free_region::FreeRegionMap;
+use rustc::middle::lang_items::UnsizeTraitLangItem;
+
+use rustc::traits::{self, ObligationCause, Reveal};
+use rustc::ty::{self, Ty, TyCtxt};
+use rustc::ty::ParameterEnvironment;
+use rustc::ty::TypeFoldable;
+use rustc::ty::subst::Subst;
+use rustc::ty::util::CopyImplementationError;
+use rustc::infer;
+
+use rustc::hir::def_id::DefId;
+use rustc::hir::map as hir_map;
+use rustc::hir::{self, ItemImpl};
+
+pub fn check<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>) {
+    if let Some(drop_trait) = tcx.lang_items.drop_trait() {
+        tcx.lookup_trait_def(drop_trait).for_each_impl(tcx, |impl_did| {
+            visit_implementation_of_drop(tcx, impl_did)
+        });
+    }
+
+    if let Some(copy_trait) = tcx.lang_items.copy_trait() {
+        tcx.lookup_trait_def(copy_trait).for_each_impl(tcx, |impl_did| {
+            visit_implementation_of_copy(tcx, impl_did)
+        });
+    }
+
+    if let Some(coerce_unsized_trait) = tcx.lang_items.coerce_unsized_trait() {
+        let unsize_trait = match tcx.lang_items.require(UnsizeTraitLangItem) {
+            Ok(id) => id,
+            Err(err) => {
+                tcx.sess.fatal(&format!("`CoerceUnsized` implementation {}", err));
+            }
+        };
+
+        tcx.lookup_trait_def(coerce_unsized_trait).for_each_impl(tcx, |impl_did| {
+            visit_implementation_of_coerce_unsized(tcx, impl_did,
+                                                   unsize_trait, coerce_unsized_trait)
+        });
+    }
+}
+
+fn visit_implementation_of_drop<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, impl_did: DefId) {
+    let items = tcx.associated_item_def_ids(impl_did);
+    if items.is_empty() {
+        // We'll error out later. For now, just don't ICE.
+        return;
+    }
+    let method_def_id = items[0];
+
+    let self_type = tcx.item_type(impl_did);
```

Comment:
Is this file all code collected from other places, or did you make changes to it? I couldn't tell.